File: src/dev/lib/kube.ts
import { checkProgram, exec } from './shell';
import { curryWrap } from './curry';

const IMAGE = 'ollama/ollama:0.1.22';
const GPU = 'NVIDIA-A10';

type PodOptions = {
  num: number,
  name: string,
  image: string,
  gpu?: string,
  command: string[],
  port?: number,
  standby?: boolean,
};

const podName = (i: number, options: { num: number, name: string }) => options.num > 1
    ? `${options.name}-pod-${i}`
    : `${options.name}-pod`;

const serviceName = (i: number, options: { num: number, name: string }) => options.num > 1
    ? `${options.name}-${i}`
    : `${options.name}`;

const webName = (i: number, options: { num: number, name: string }) => options.num > 1
    ? `${options.name}-${i}-dcct.nrp-nautilus.io`
    : `${options.name}-dcct.nrp-nautilus.io`;

const pod = (i:number, options: PodOptions) => {
  const name = podName(i, options);
  return {
    apiVersion: 'v1',
    kind: 'Pod',
    metadata: {
      name,
      labels: { 'k8s-app': name }
    },
    spec: {
      containers: [
        {
          image: options.image,
          name: name,
          command: options.command || undefined,
          // command: ['/bin/sh', '-c', 'sleep 30'],
          resources: {
            limits: {
              cpu: '2',
              memory: '8Gi',
              'nvidia.com/gpu': '1'
            }
          },
          env: [
            {
              name: 'PATH',
              value: '/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
            },
            {
              name: 'LD_LIBRARY_PATH',
              value: '/usr/local/nvidia/lib:/usr/local/nvidia/lib64'
            },
            {
              name: 'NVIDIA_DRIVER_CAPABILITIES',
              value: 'compute,utility'
            }
          ],
        }
      ],
      affinity: options.gpu ? {
        nodeAffinity: {
          requiredDuringSchedulingIgnoredDuringExecution: {
            nodeSelectorTerms: [{
                matchExpressions: [{
                    key: 'nvidia.com/gpu.product',
                    operator: 'In',
                    values: [ options.gpu ]
                  }
                ]}
            ]}
        }
      } : undefined
    }
  };
};

const service = (i:number, options: PodOptions) => {
  const name = serviceName(i, options);
  return {
    apiVersion: 'v1',
    kind: 'Service',
    metadata: {
      name,
      labels: { 'k8s-app': name }
    },
    spec: {
      type: 'ClusterIP',
      ports: [{ port: 80, targetPort: options.port || 80 }],
      selector: { 'k8s-app': podName(i, options) }
    }
  };
};

const ingress = (i:number, options: PodOptions) => {
  const name = serviceName(i, options);
  return {
    apiVersion: 'networking.k8s.io/v1',
    kind: 'Ingress',
    metadata: {
      name,
      annotations: { 'kubernetes.io/ingress.class': 'haproxy' }
    },
    spec: {
      rules: [{
        host: webName(i, options),
        http: {
          paths: [{
            path: '/',
            pathType: 'ImplementationSpecific',
            backend: {
              service: {
                name,
                port: { number: 80 }
              }
            }
          }]
        }
      }],
      tls: [{ hosts: [webName(i, options)] }]
    }
  };
};

const getPodsStatus = async (pods: PodOptions) => {
  const response = await exec`kubectl get pod`;

  const res = response
    .trim()
    .split('\n')
    .map(line => line.split(/\s+/))
    .filter(podName => podName[0].includes(pods.name))
    .map(podName => podName[2]);

  if (res.length === 0)
    return Array(pods.num).fill('NotFound');
  return res;
}

const waitPodUntilReady = async (i, pods: PodOptions) => {
  while (true) {
    const podsStatus = await getPodsStatus(pods);
    if (podsStatus[i] === 'NotFound')
      return Promise.reject('Pod not found');
    if (podsStatus[i] === 'Running')
      return;
    await new Promise(resolve => setTimeout(resolve, 3000));
  }
}

const randUUID = () =>
  'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });

const checkAvailability = async (options: PodOptions) => {
  const response = await fetch(
    'https://portal.nrp-nautilus.io/rpc',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'guest.ListNodeInfo',
        params: {},
        // random uuid
        id: randUUID()
      })
    }
  );
  const json = await response.json();
  if (json.error)
    return Promise.reject(json.error);
  
  return json.result.Nodes
    .filter(node => options.gpu ? node.GPUType === options.gpu : true)
    .filter(node => node.Taints.length === 0)
    .map(node => +(node.GPUAvailable))
    .reduce((a, b) => a + b, 0);
}

const upKubernetes = (options: PodOptions) => async () => {
  console.log('Preparing kubernetes resources');

  const available = await checkAvailability(options);
  if (available < options.num)
    return Promise.reject('Not enough resources');

  const podsStatus = await getPodsStatus(options);
  await Promise.all(
    podsStatus.map(async (status, i) => {
      const pod_i = pod(i, options);
      const pod_name = podName(i, options);

      if (status !== 'Running') {
        if (status !== 'NotFound') {
          await exec`kubectl delete pod ${pod_name}`;
        }
        if (status !== 'Pending' || status !== 'Waiting') {
          console.log(`Creating pod ${pod_name}`);
          const pod_script = new Response(JSON.stringify(pod_i));
          await exec`kubectl create -f - < ${pod_script}`;
        }
        console.log(`Waiting ${pod_name}`);
        await waitPodUntilReady(i, options);

        const service_i = service(i, options);
        const service_script = new Response(JSON.stringify(service_i));
        await exec`kubectl create -f - < ${service_script}`;

        const ingress_i = ingress(i, options);
        const ingress_script = new Response(JSON.stringify(ingress_i));
        await exec`kubectl create -f - < ${ingress_script}`;

        console.log('Cold start, waiting for 10 seconds');
        await new Promise(resolve => setTimeout(resolve, 10000));
      }
      console.log(`Pod ready: ${pod_name}`);
    })
  );
}

const downKubernetes = (options: PodOptions) => async () => {
  if (options.standby) {
    console.log('Standby mode, skipping cleanup');
    return;
  }
    
  console.log('Cleaning up kubernetes resources');
  await Promise.all(
    Array.from({ length: num }, (_, i) => i + 1)
    .map(async i => {
      const name = podName(i, options);
      console.log(`Deleting pod ${name}`);
      await exec`kubectl delete pod ${name}`;
    })
  );
}

const useKubernetes = curryWrap(
  upKubernetes,
  downKubernetes
);

export {
  useKubernetes,
  webName,
};

File: src/dev/lib/kube.py
import json
import os
import subprocess
from time import sleep
from uuid import uuid4

import requests


def gpu_keys(gpu):
    # https://docs.nationalresearchplatform.org/userdocs/running/gpu-pods/#requesting-high-end-gpus
    high_end = {
        "NVIDIA-A100": "nvidia.com/a100",
        "NVIDIA-A40": "nvidia.com/a40",
        "NVIDIA-RTX-A6000": "nvidia.com/rtxa6000",
        "Quadro-RTX-8000": "nvidia.com/rtx8000",
        "NVIDIA-GH200-480GB": "nvidia.com/gh200",
    }
    if gpu in high_end:
        return {
            high_end[gpu]: "1",
        }
    return {
        "nvidia.com/gpu": "1",
    }


class Pod:
    def __init__(self, options):
        self.options = options
        self.name = f"{options['name']}-pod"
        self.json = pod_json(options, self.name)

    def status(self):
        res = (
            subprocess.run(
                [
                    "kubectl",
                    "get",
                    "pod",
                    self.name,
                    '--template={{printf "%s" .status.phase}}',
                ],
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
            )
            .stdout.decode("utf-8")
            .strip()
        )
        return None if res == "" else res

    def check_spec(self):
        response = requests.post(
            "https://portal.nrp-nautilus.io/rpc",
            json={
                "jsonrpc": "2.0",
                "method": "guest.ListNodeInfo",
                "params": {},
                "id": str(uuid4()),
            },
        )
        print(response)
        json = response.json()
        if "error" in json:
            raise Exception(json["error"]["message"])
        gpu_needs = list(
            filter(
                lambda n: "gpu" not in self.options
                or n["GPUType"] == self.options["gpu"],
                json["result"]["Nodes"],
            )
        )
        # print(list(map(lambda n: n['Taints'], gpu_needs)))
        # no_taint_nodes = list(filter(
        #     lambda n: n['Taints'] is None or n['Taints'] == [], gpu_needs))
        gpus = list(map(lambda n: int(n["GPUAvailable"]), gpu_needs))
        return sum(gpus)

    def up(self):
        print("Preparing kubernetes resources")
        # available = self.check_spec()
        # if available == 0:
        #     raise Exception(f'Not enough resources: {self.options["gpu"]}')

        status = self.status()
        if status != "Running":
            if status in ["Completed", "Failed", "Error"]:
                subprocess.run(["kubectl", "delete", "pod", self.name])
                while self.status() is not None:
                    sleep(1)
            if status not in ["Pending", "ContainerCreating", "Waiting"]:
                print(f"Creating {self.name}")
                subprocess.run(
                    ["kubectl", "create", "-f", "-"], input=self.json, text=True
                )

            print(f"Waiting for {self.name} to be ready")
            while self.status() != "Running":
                sleep(1)

        print(f"Pod ready: {self.name}")

    def exec_code(self, *command):
        res = subprocess.run(
            ["kubectl", "exec", self.name, "--", *command],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
        )
        return res.returncode

    def port_forward(self, port):
        id = os.fork()
        if id == 0:
            res = subprocess.run(
                ["kubectl", "port-forward", f"pod/{self.name}", f"{port}"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        else:
            # while requests.get(f'http://localhost:{port}').status_code != 200:
            while True:
                try:
                    requests.get(f"http://localhost:{port}")
                    break
                except requests.exceptions.ConnectionError:
                    sleep(1)
                except requests.exceptions.Timeout:
                    print("Timeout")
                    pass
            print(f"Server accessible at http://localhost:{port}")
            os.kill(id, 9)

    def down(self):
        if self.status() is not None and not self.options["standby"]:
            print(f"Deleting {self.name}")
            subprocess.run(["kubectl", "delete", "pod", self.name])
            while self.status() is not None:
                sleep(1)


def pod_json(options, name):
    return json.dumps(
        {
            "apiVersion": "v1",
            "kind": "Pod",
            "metadata": {
                "name": name,
                "labels": {
                    "k8s-app": name,
                },
            },
            "spec": {
                "containers": [
                    {
                        "name": name,
                        "image": options["image"],
                        "command": []
                        if "command" not in options
                        else options["command"],
                        "resources": {
                            "limits": {
                                "cpu": "2",
                                "memory": "8Gi",
                                "ephemeral-storage": "64Gi",
                            }
                            | gpu_keys(options["gpu"])
                            if "gpu" in options
                            else {},
                        },
                        "volumeMounts": [
                            {
                                "mountPath": "/models",
                                "name": "data",
                            }
                        ]
                        if "data" in options
                        else [],
                        "env": [
                            {
                                "name": "PATH",
                                "value": "/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                            },
                            {
                                "name": "LD_LIBRARY_PATH",
                                "value": "/usr/local/nvidia/lib:/usr/local/nvidia/lib64",
                            },
                            {
                                "name": "NVIDIA_DRIVER_CAPABILITIES",
                                "value": "compute,utility",
                            },
                        ],
                    },
                ],
                "volumes": [
                    {
                        "name": "data",
                        "persistentVolumeClaim": {"claimName": "ollama-models"},
                    }
                ],
                "affinity": {
                    "nodeAffinity": {
                        "requiredDuringSchedulingIgnoredDuringExecution": {
                            "nodeSelectorTerms": [
                                {
                                    "matchExpressions": [
                                        {
                                            "key": "nvidia.com/gpu.product",
                                            "operator": "In",
                                            "values": [options["gpu"]],
                                        },
                                    ],
                                },
                            ],
                        },
                    },
                }
                if "gpu" in options
                else {},
            },
        }
    )


def deploy_from_yml(filename):
    subprocess.Popen(f"kubectl apply -f {filename}", shell=True).communicate()
    # get deployment names from file
    deployments = (
        subprocess.check_output(
            f"kubectl get -f {filename} -o name | grep deployment", shell=True
        )
        .decode()
        .strip()
        .split("\n")
    )

    watch_process = subprocess.Popen(
        f"kubectl events deployment --no-headers -w", shell=True, stdout=subprocess.PIPE
    )

    wait_process = subprocess.Popen(
        f"kubectl wait --for=condition=available {
            ' '.join(deployments)
        } --timeout=-1s && kill {watch_process.pid}",
        shell=True,
    )

    sleep(2)
    while True:
        output = watch_process.stdout.readline()
        if watch_process.poll() is not None:
            break
        if output:
            print(output.decode().strip())


def port_forward(service, port):
    id = os.fork()
    if id == 0:
        res = subprocess.run(
            f"kubectl port-forward service/{service} --pod-running-timeout=1h {port}:80",
            shell=True,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    else:
        while True:
            try:
                requests.get(f"http://localhost:{port}")
                break
            except requests.exceptions.ConnectionError:
                sleep(1)
            except requests.exceptions.Timeout:
                print("Timeout")
                pass
        print(f"Server accessible at http://localhost:{port}")
        os.kill(id, 9)


def delete_deployment():
    print(f"Deleting all deployments")
    subprocess.run(["kubectl", "delete", "deployment"])
